[Lye Wen Jun] iP

xlzior opened the pull request:
# Iris 🤖

> Who needs _Siri_ when you have **Iris**?!

## Features

- simple, text-based interface
- written in Java: _write once, run anywhere_

## How to use

1. Create a todo with `todo [task name]`
2. View all tasks with `list`
3. Mark the todo as complete with `done 1`
4. Delete the completed todo with `delete 1`

## Developer Experience

Iris is easily extendable -- simply create a new class in the `command` package which extends the `Command` base class, and implement `runSilently(TaskList tasks)` and `say(TaskList tasks, Ui ui)`.

```java
public class FindCommand extends Command {
    private String searchTerm;
    private List<Task> searchResults;

    public FindCommand(String searchTerm) {
        this.searchTerm = searchTerm;
    }

    @Override
    public void runSilently(TaskList tasks) throws IrisException {
        this.searchResults = tasks.find(this.searchTerm);
    }

    @Override
    public void say(TaskList tasks, Ui ui) {
        for (int i = 0; i < searchResults.size(); i++) {
            ui.say(String.format("%d. %s", i + 1, searchResults.get(i)), i == 0);
        }
    }
}
```

### Tasks

- [x] Add different types of tasks (todos, deadlines, events) to your todo list
- [X] Manage tasks in your todo list by marking them as complete or deleting them
- [X] Find tasks in your todo list by searching for them
- [ ] Access Iris through a GUI

Original repository: [link](https://github.com/xlzior/ip)

Commit:
Add Gradle support

xlzior committed:
Level 1: Implement skeletal version

* greet
* echo
* exit

xlzior committed:
Level 2: Implement add and list

* add
* list

xlzior committed:
Level 3: Convert Task from a String to a class

xlzior committed:
Level 3: Implement mark as done

xlzior committed:
Level 4: Implement different task types as subclasses of Task

xlzior committed:
Level 4: Refactor add function

xlzior committed:
Level 4: Implement deadline by

xlzior committed:
Level 4: Implement event at

xlzior committed:
Level 4: Implement list status describing number of incomplete tasks in list

xlzior committed:
A-TextUiTesting: Implement semi-automated testing

xlzior committed:
Level 5: Handle unknown command exception

xlzior committed:
Level 5: Handle todo empty description exception

xlzior committed:
Level 5: Handle other command empty description exception

xlzior committed:
Level 6: Implement delete

xlzior committed:
Level 5: Handle more exceptions

xlzior committed:
Level 7: Implement loading taskList from file

xlzior committed:
Level 7: Implement saving commands to taskList file

xlzior committed:
Level 8: Parse date and time information into LocalDate objects

xlzior committed:
Merge branch 'branch-Level-7'

xlzior committed:
Merge branch 'branch-Level-8'

xlzior committed:
A-MoreOOP: Split functions into separate files

xlzior committed:
Level-8: Update test cases with new date format requirement

xlzior committed:
A-MoreOOP: Convert Storage from static to non-static

xlzior committed:
A-MoreOOP: Convert TaskList from static to non-static

xlzior committed:
A-MoreOOP: Convert Ui from static to non-static

xlzior committed:
A-Packages: Divide classes into packages

xlzior committed:
A-Packages: Fix semi-automated testing script

xlzior committed:
A-JUnit: Add JUnit tests for duke.task

xlzior committed:
Level 7: Create dedicated data folder

xlzior committed:
A-Jar: Package the app as a JAR file

xlzior committed:
A-JavaDoc: Add JavaDoc comments

xlzior committed:
A-JavaDoc: Fix JavaDoc tenses

xlzior committed:
A-JavaDoc: Fix JavaDoc line breaks

xlzior committed:
A-CodingStandard: Tweak code to comply with coding standard

xlzior committed:
Level-9: Implement finding tasks by keyword

xlzior committed:
Merge branch 'branch-A-JavaDoc'

xlzior committed:
Merge branch 'branch-A-CodingStandard'

xlzior committed:
Merge branch 'branch-Level-9'

xlzior committed:
Level-7: Delete stored tasks from previous run in runtest.sh

xlzior committed:
Rename Duke to Iris

xlzior committed:
Add data/tasks.txt to .gitignore

xlzior committed:
A-Packages: Extract out command package

xlzior committed:
A-MoreOOP: Refactor out greeting and farewell

xlzior committed:
A-MoreOOP: Refactor out ByeCommand

xlzior committed:
Remove unused imports

xlzior committed:
Add missing JavaDocs details

xlzior committed:
Level 5: Fix typo in exceptions

xlzior committed:
Merge branch 'add-gradle-support'

xlzior committed:
A-CheckStyle: Set up checkstyle

xlzior committed:
A-CheckStyle: Fix checkstyle violations

xlzior committed:
Level-10: Set up javafx GUI skeleton

xlzior committed:
Level 10: Rewrite Command interface

Previously used `runSilently` and `say` to split between execution of the action and UI response respectively

Now uses `run` for execution and `store` for writing to Storage, if necessary

xlzior committed:
Level 10: Integrate Iris responses into GUI

xlzior committed:
Level-10: Update GUI aesthetics

xlzior committed:
Level 10: Update GUI JavaDocs

xlzior committed:
Merge branch 'branch-Level-10'

clyveycui commented on src/main/java/iris/Parser.java:
```diff
@@ -0,0 +1,74 @@
+     * @throws IrisException if command is invalid
+     */
+    public static Command parse(String command) throws IrisException {
+        if (command.equals("list")) {
+            return new ListCommand();
+        } else if (command.startsWith("todo")) {
+            return new ToDoCommand(getMetadata(command));
+        } else if (command.startsWith("deadline")) {
+            String[] splitted = getMetadata(command).split(" /by ");
+            if (splitted.length != 2) {
+                throw new IrisException("deadline should have 2 arguments: a name and a time");
+            }
+            return new DeadlineCommand(splitted[0], splitted[1]);
+        } else if (command.startsWith("event")) {
+            String[] splitted = getMetadata(command).split(" /at ");
+            if (splitted.length != 2) {
+                throw new IrisException("event should have 2 arguments: a name and a time");
+            }
+            return new EventCommand(splitted[0], splitted[1]);
+        } else if (command.startsWith("done")) {
+            int index = parseInt(getMetadata(command));
+            return new DoneCommand(index);
+        } else if (command.startsWith("delete")) {
+            int index = parseInt(getMetadata(command));
+            return new DeleteCommand(index);
+        } else if (command.startsWith("find")) {
+            String searchTerm = getMetadata(command);
+            return new FindCommand(searchTerm);
+        } else if (command.startsWith("bye")) {
+            return new ByeCommand();
+        } else {
+            throw new IrisException("I'm sorry, but I don't know what that means.");
+        }
```
Using switch case might make it look neater, but that's just my personal opinion.

xlzior replied:
tbh I did if/else on purpose because I was worried they would add a new requirement for commands which aren't a single word, but yeah I think I can change it to switch/case now. thank you!

clyveycui commented on src/main/java/iris/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+     * @param index index of task to get
+     * @return Task at the given index
+     */
+    public Task get(int index) {
+        return tasks.get(index);
+    }
```
Naming it getTask might make it's usage clearer, but this is fine too.

clyveycui commented on src/main/java/iris/task/Deadline.java:
```diff
@@ -0,0 +1,60 @@
+     * @param index index of this Task in the TaskList
+     * @return String representing the command to re-create this Deadline object
+     */
+    @Override
+    public String toCommand(int index) {
+        return String.format(
+                "deadline %s /by %s\n%s",
+                this.name,
+                this.by,
+                super.toCommand(index)
+        );
+    }
```
Pretty cool design here, first one that I have seen using a toCommand method for each individual task object.

clyveycui reviewed (commented):
Awesome, just a few minor nitpicks, but lgtm.

Amadeus-Winarto commented on src/main/java/iris/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+     * @param index index of Task to lookup
+     * @throws IrisException for invalid Task index e.g. too small or too large
+     */
+    private void validateTaskIndex(int index) throws IrisException {
+        if (index <= 0) {
+            throw new IrisException("Please enter a valid task index.");
+        }
+        int count = tasks.size();
+        if (index > count) {
+            throw new IrisException(String.format("Your task list only has %d items", count));
+        }
+    }
```
I like this explicit validation method

Amadeus-Winarto commented on src/test/java/iris/DeadlineTest.java:
```diff
@@ -0,0 +1,46 @@
+
+import iris.task.Deadline;
+
+public class DeadlineTest {
```
General question (and may have missed it also), but will you be doing a test on storage? Like if it successfully read/write to storage file?

Amadeus-Winarto reviewed (commented):
Great job! Nice design too!

xlzior committed:
Add assertions to assert that tasks is not null

xlzior committed:
A-CodeQuality: Convert if statement to switch statement

xlzior committed:
A-CodeQuality: Fill in empty catch blocks for IOException

xlzior committed:
A-CodeQuality: Replace magic literal with named constant

xlzior committed:
A-CodeQuality: Use Java enums to make the code obvious

xlzior committed:
Merge pull request #1 from xlzior/branch-A-Assertions

Use assertions

xlzior committed:
Merge branch 'master' into branch-A-CodeQuality

xlzior committed:
Merge pull request #2 from xlzior/branch-A-CodeQuality

Improve code quality

xlzior committed:
C-Priority: Implement task priorities

Tasks do not have a priority level

Some tasks may be more important than others

Let's allow the user to specify that some tasks are more important than others using the '!' modifier to the existing
task commands (todo, deadline, event). The new commands are todo!, todo!! etc.

Exclamation marks convey a sense of urgency/importance appropriate of high priority tasks.

Task priorities are represented by the TaskPriority enumeration

xlzior committed:
Merge pull request #3 from xlzior/branch-C-Priority

C-Priority: Implement task priorities

xlzior committed:
A-UserGuide: Add a user guide

xlzior committed:
Set theme jekyll-theme-hacker

xlzior committed:
Set theme jekyll-theme-cayman

xlzior committed:
Set theme jekyll-theme-hacker

xlzior committed:
Set theme jekyll-theme-cayman

xlzior committed:
A-UserGuide: Update user guide with priority

xlzior committed:
Implement closing app on bye command

xlzior committed:
Update invalid date error message to be more informative

xlzior committed:
A-CodeQuality: Convert constant names to uppercase

xlzior committed:
A-CodeQuality: Fix import order

xlzior committed:
Update list command to be more informative when task list is empty
